datafusion/physical-optimizer/src/pushdown_sort.rs

```diff
@@ -0,0 +1,128 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Sort Pushdown Optimization
+//!
+//! This optimizer attempts to push sort requirements down through the execution plan
+//! tree to data sources that can natively handle them (e.g., by scanning files in
+//! reverse order).
+//!
+//! ## How it works
+//!
+//! 1. Detects `SortExec` nodes in the plan
+//! 2. Calls `try_pushdown_sort()` on the input to recursively push the sort requirement
+//! 3. Each node type defines its own pushdown behavior:
+//!    - **Transparent nodes** (CoalesceBatchesExec, RepartitionExec, etc.) delegate to
+//!      their children and wrap the result
+//!    - **Data sources** (DataSourceExec) check if they can optimize for the ordering
+//!    - **Blocking nodes** return `Unsupported` to stop pushdown
+//! 4. Based on the result:
+//!    - `Exact`: Remove the Sort operator (data source guarantees perfect ordering)
+//!    - `Inexact`: Keep Sort but use optimized input (enables early termination for TopK)
+//!    - `Unsupported`: No change
+//!
+//! ## Current capabilities (Phase 1)
```

Comment:
If we are going to leave references to phase 2 in the code, I think we should file a ticket and leave a link to that new ticket for future readers. If we leave it as a comment it will be harder to find and track the current status of I fear

Reply:
Created the ticket:
https://github.com/apache/datafusion/issues/19329